README.md
# libserial README.md

This is a library for interfacing with serial ports written in C++. Under the hood the termios library is used,
so the library should compile without problems on any POSIX-compliant system.

    Modern C++.
    RAII compliant.
    No dependencies.
    Only two files (libserial.h and libserial.cpp).

The library has been tested on Linux and used in a number of embedded systems.

# Compiling

# Examples

Tests programs in the test directory are also examples of usage.


doc/changelog
/**
 * libserial
 * Manejo de comunicaciones por puerto serie
 *
 * @file      changelog
 * @brief     Historial de revisiones
 * @author    José Luis Sánchez Arroyo
 * @date      2020.05.22
 * @version   1.7
 *
 * Copyright (c) 2005-2020 José Luis Sánchez Arroyo
 * This software is distributed under the terms of the LGPL version 2 and comes WITHOUT ANY WARRANTY.
 * Please read the file COPYING.LIB for further details.
 */
-----------------------------------------------------------------------------------------------------------------------
1.7 (2020.05.22)
Añadida la función WaitSend().
Corregido un error en PendingWrite: si solo quedaba el byte del shift register, se devolvía 0.

1.6.0 (2019.02.04)
Modificación del interfaz de la clase:
- Los tipos de puntero de Read y Write ahora son punteros a void. Fin de la pesadilla.
Bugfix: El timeout de lectura no se reinicia por la llegada de algún carácter.

1.5.0 (2019.02.01)
Modificación del interfaz de la clase:
- El constructor ahora no admite parámetros.
- La función 'Init' es renombrada a 'Open', que admite nuevos parámetros para especificar el
  dispositivo y la modalidad de apertura (bloqueante o no bloqueante).
- La velocidad (baudrate) se especifica como valor entero y se convierte internamente.
- Ampliada la tabla de conversión de velocidades para incluir las nuevas constantes (hasta 4 Mbps)
- Nueva función Close para cerrar el dispositivo.
- Nueva función SetBlocking para definir el modo de bloqueo en lectura.
- IsValid pasa a llamarse IsOpen. El objeto también es válido cuando el puerto está cerrado.
- Eliminadas las funciones IsTxFIFOempty (redundante) y GetBaudValue (innecesaria).
- Las funciones PendingRead y PendingWrite ahora devuelven el número de bytes pendientes.
Actualización de contenido y estilo de la documentación.
Limpieza de código muerto o inútil.

1.4.0 (2017.05.09)
La función Init recibe ahora parámetros enumerados, en lugar de referir a constantes de termios.h

1.3.3 (2016.06.03)
Ampliación de la funcionalidad del programa de pruebas.
Nueva función GetBaudValue, reverso de GetBaudCode.
Ambas funciones pasan a ser estáticas.

1.3.2 (2016.06.01)
El programa de test ahora admite que el mensaje a enviar sea opcional y el tiempo de espera infinito.

1.3.1 (2016.05.27)
Añadido un programa de test.

1.3.0 (2016.05.03)
Nuevas funciones PendingRead y PendingWrite para determinar si quedan datos en los buffers de entrada y salida.

1.2.3 (2016.04.27)
Eliminadas un par de trazas que andaban por ahí perdidas.
Actualización de la documentación.

1.2.2 (2016.04.01)
La identificación de versión se segrega a un fichero fuente independiente version.cpp

1.2.1 (2016.03.09)
Modificaciones por el cambio de interfaz del módulo de gestión de versiones.

1.2.0 (2014.04.23)
Reordenación de código, moviendo las cabeceras al directorio raíz del módulo.
Información para doxygen movida a las cabeceras.

1.1.4 (2008.07.15)
Se incluye función IsTxFIFOempty para comprobar si la UART serie transmitió todos los datos del buffer.

1.1.3 (2008.02.15)
Corrección al fijar la velocidad, a lo POSIX.

1.1.2 (2008.01.24)
Añadida la función WriteByte para escribir un solo byte.

1.1.1 (2007.11.26)
La función GetVersion no era estática como debería ser.
Eliminado el prototipo de la función Init antigua, que no debería estar ahí.

1.1.0 (2007.10.24)
Modificada la función Init para admitir parámetros en un formato más "coherente".
Añadida la función GetBaudCode que convierte valores de velocidad en códigos para Init.

1.0.0 (2006.10.24)
Actualización de la documentación.
Cambios estéticos.
Release 1.

0.2.1 (2006.01.30)
Reemplazado gettimeofday por TimerClass.

0.2.0 (2006.01.23)
La librería comienza su vida independiente del módulo tci.


test/test_libserial.cpp
/**
 * libserial
 * Manejo de comunicaciones por puerto serie
 *
 * @file      test_libserial.cpp
 * @brief     Pruebas unitarias
 * @author    José Luis Sánchez Arroyo
 * @date      2019.02.01
 * @version   1.5.0
 *
 * Copyright (c) 2005-2016 José Luis Sánchez Arroyo
 * This software is distributed under the terms of the LGPL version 2 and comes WITHOUT ANY WARRANTY.
 * Please read the file COPYING.LIB for further details.
 */

#include "libserial.h"
#include <libUtility/tracelog.h>
#include <libUtility/versioninfo.h>
#include <libUtility/timer.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include <vector>

MainVersion(test_libserial, 1.5.0)                      //!< Nombre y versión del programa

/**--------------------------------------------------------------------------------------------------
 * @brief       Constantes de configuración
 * ------ */
const char* DEFAULT_SERIAL_DEV = "/dev/ttyS0";          //!< Dispositivo serie por omisión
uint32_t    DEFAULT_SERIAL_BPS = 9600;                  //!< Velocidad de conexión por omisión
uint32_t    DEFAULT_SERIAL_TIMEOUT = 1000;              //!< Tiempo de espera máximo por omisión
uint32_t    MAX_MESSAGE_LEN = 512;                      //!< Longitud máxima del mensaje en bytes

/**--------------------------------------------------------------------------------------------------
 * @brief       Leer datos del puerto serie
 * ------*/
void ReadData(Serial& port, uint32_t timeout)
{
  Timer timer;
  uint64_t start_time = timer.GetTimeNow();
  uint32_t received = 0;
  std::cout << "Reading..." << std::endl;
  for (timer.SetAlarm(timeout); timer.IsExpired() == false; )
  {
    uint8_t byte;
    if (port.Read(&byte, 1, timer.GetRemain()) == false)
      break;
    printf("%09ld %3d  %02X\n", timer.GetTimeNow() - start_time, ++received, byte);
  }
  printf("%09ld Terminado - %d bytes recibidos.\n", timer.GetTimeNow() - start_time, received);
}

/**--------------------------------------------------------------------------------------------------
 * @brief       Leer paquete de datos del puerto serie
 * ------*/
void ListenData(Serial& port, uint32_t timeout)
{
  Timer timer;
  uint64_t start_time = timer.GetTimeNow();
  uint32_t received = 0;
  std::cout << "Listening..." << std::endl;
  while (1)
  {
    uint8_t byte;
    if (port.Read(&byte, 1, timeout) == true)
      printf("%09ld %3d  %02X\n", timer.GetTimeNow() - start_time, ++received, byte);
    else if (received > 0)
      break;
  }
  printf("%09ld Terminado - %d bytes recibidos.\n", timer.GetTimeNow() - start_time, received);
}

/**--------------------------------------------------------------------------------------------------
 * @brief       Escribir datos a puerto serie
 * ------*/
void WriteData(Serial& port, std::vector<const char*>& bytes)
{
  if (bytes.size() == 0)
  {
    std::cout << "Mensaje vacío\n";
    return;
  }
  if (bytes.size() > MAX_MESSAGE_LEN)
  {
    std::cout << "Mensaje demasiado largo, máximo " << MAX_MESSAGE_LEN << " bytes.\n";
    return;
  }
  uint8_t message[MAX_MESSAGE_LEN];
  uint32_t msg_len = 0;
  for (unsigned ix = 0; ix < bytes.size(); ++ix)
  {
    uint32_t byte = strtoul(bytes[ix], 0, 16);
    if (byte > 0xff)
    {
      std::cout << "Byte incorrecto '" << bytes[ix] << "'\n";
      return;
    }
    message[msg_len++] = byte;
  }

  std::cout << "Enviando mensaje: ";
  for (unsigned ix = 0; ix < msg_len; ++ix)
    printf("%02X ", message[ix]);
  std::cout << std::endl;
  if (port.Write(message, msg_len) == false)
    std::cout << "Error durante el envío." << std::endl;
  std::cout << "Mensaje enviado.\n";
}

/**--------------------------------------------------------------------------------------------------
 * @brief       Consola para enviar y recibir datos
 * ------*/
void Console(Serial& port)
{
  for (bool loop = true; loop; )
  {
    char line[MAX_MESSAGE_LEN];

    std::cout << "[TestSerial (? para ayuda)]: ";
    std::cin.getline(line, MAX_MESSAGE_LEN - 1);
    strtok(line, " ");
    switch (line[0])
    {
      case 0:
      case '?':                                 // Help!
        std::cout << 
          "(w)rite <message>    Enviar datos al puerto serie\n"
          "   message:          Pares de dígitos hexadecimales separados por espacio\n"
          "(r)ead  <time>       Recibir datos del puerto serie\n"
          "   time:             Tiempo máximo de recepción, en milisegundos\n"
          "(c)ommand <time> <message>  Enviar datos al puerto y esperar respuesta inmediatamente\n"
          "   time:             Tiempo máximo de recepción, en milisegundos\n"
          "   message:          Mensaje a enviar (igual que en (w)rite)\n"
          "(a)nswer  <time> <message>  Esperar indefinidamente un mensaje y, tras recibirlo, enviar respuesta\n"
          "   time:             Tiempo máximo de espera entre bytes del mensaje, en milisegundos\n"
          "   message:          Mensaje a enviar (igual que en (w)rite)\n"
          "(l)isten <timeout>   Esperar indefinidamente datos desde el puerto\n"
          "   timeout:          Una vez leído un byte, tiempo máximo de espera por un byte\n"
          "(q)uit: Salir de la consola\n"
          ;
        break;

      case 'q':                                 // Quit
        std::cout << "Saliendo de la consola.\n";
        loop = false;
        break;

      case 'r':                                 // Read
      case 'l':                                 // Listen
      {
        uint32_t timeout = 0;
        char* t_out = strtok(0, " ");
        if (t_out)
          timeout = strtoul(t_out, 0, 0);
        if (line[0] == 'r')
          ReadData(port, timeout);
        else
          ListenData(port, timeout);
        break;
      }

      case 'w':                                 // Write
      {
        std::vector<const char*> bytes;
        for (char* ptr = strtok(0, " "); ptr; ptr = strtok(0, " "))
          bytes.push_back(ptr);
        if (bytes.size() > 0)
          WriteData(port, bytes);
        break;
      }

      case 'c':                                 // Command (write, then read)
      case 'a':                                 // Answer (read, then write)
      {
        uint32_t timeout = 0;
        char* t_out = strtok(0, " ");
        if (t_out)
          timeout = strtoul(t_out, 0, 0);
        std::vector<const char*> bytes;
        for (char* ptr = strtok(0, " "); ptr; ptr = strtok(0, " "))
          bytes.push_back(ptr);

        if (line[0] == 'a')
          ListenData(port, timeout);

        if (bytes.size() > 0)
          WriteData(port, bytes);

        if (line[0] == 'c')
          ReadData(port, timeout);
        break;
      }
      default:
        std::cout << "Nonsense.\n";
        break;
    }
  }
}

/**--------------------------------------------------------------------------------------------------
 * @brief       Salida con mensaje de sintaxis
 * ------ */
void Abort(const char* prog)
{
  std::cout << prog << " : Pruebas de transferencia de datos por puerto serie\n"
    "Sintaxis: " << prog << " -h | [-d <device>] [-s <speed> [-t <timeout>] [-i | <mensaje>]\n"
    "  -h : Esta ayuda\n"
    "  -d : Dispositivo de conexión [/dev/ttyS0]\n"
    "       <device>  : Path del dispositivo serie a utilizar\n"
    "  -s : Velocidad de conexión [9600]\n"
    "       <speed>   : Velocidad de conexión requerida (9600, 19200, 38400...)\n"
    "  -t : Tiempo máximo de espera en lectura\n"
    "       <timeout> : Tiempo de espera, en milisegundos\n"
    "  -i : Modo interactivo\n"
    "mensaje: secuencia de bytes a enviar por el puerto (pares de dígitos hexadecimales separados por espacio)\n"
    ;
  exit(-1);
}

/**--------------------------------------------------------------------------------------------------
 * @brief       main
 * ------*/
int main(int argc, char* argv[])
{
  VersionInfo::Show();

  /*--- Procesamiento de línea de comandos ---*/
  const char* serial_dev = DEFAULT_SERIAL_DEV;
  uint32_t serial_bps = DEFAULT_SERIAL_BPS;
  uint32_t serial_timeout = DEFAULT_SERIAL_TIMEOUT;
  bool interactive = false;
  std::vector<const char*> bytes;

  for (int param = 1; param < argc; param++)
  {
    if (argv[param][0] == '-')
    {
      switch (argv[param][1])                           // Parámetros que empiezan por '-'
      {
        case 'h':
          Abort(argv[0]);
          break;                                        // no hace falta...

        case 'd':                                       // Dispositivo de conexión (puerto serie)
          if (param + 1 < argc)
            serial_dev = argv[++param];
          break;

        case 's':                                       // Velocidad de conexión
          if (param + 1 < argc)
            serial_bps = strtoul(argv[++param], 0, 0);
          break;

        case 't':                                       // Timeout de lectura en milisegundos
          if (param + 1 < argc)
            serial_timeout = strtoul(argv[++param], 0, 0);
          break;

        case 'i':
          if (bytes.size() > 0)
            Logger(Log::Error) << "Especifique modo interactivo o mensaje en línea, no ambos." << Log::end;
          interactive = true;
          break;

        default:                                        // wrong
          Logger(Log::Error) << "Parámetro incorrecto: '" << argv[param] << "'" << Log::end;
          break;
      }
    }
    else
    {
      if (interactive)
        Logger(Log::Error) << "Especifique modo interactivo o mensaje en línea, no ambos." << Log::end;
      bytes.push_back(argv[param]);
    }
  }
  std::cout << "Parámetros: device='" << serial_dev << "'  speed=" << serial_bps << "  timeout=" << serial_timeout << std::endl;
  if (bytes.size() > 0)
  {
    std::cout << "Mensaje   : ";
    for (unsigned ix = 0; ix < bytes.size(); ++ix)
      printf("%s ", bytes[ix]);
    std::cout << std::endl;
  }

  /*--- Apertura del puerto serie ---*/
  Serial port;
  if (port.Open(serial_dev, serial_bps) == false)
    Logger(Log::Error) << "Error abriendo / configurando el puerto serie" << Log::end;

  /*--- Enviar y recibir mensaje ---*/
  if (interactive)
    Console(port);
  else
  {
    if (bytes.size() > 0)
      WriteData(port, bytes);
    ReadData(port, serial_timeout);
  }
}


include/libserial.h
/**
 * libserial
 * Manejo de comunicaciones por puerto serie
 *
 * @file      libserial.h
 * @brief     Clase de manejo del puerto serie
 * @author    José Luis Sánchez Arroyo
 * @date      2020.05.22
 * @version   1.7
 *
 * Copyright (c) 2005-2020 José Luis Sánchez Arroyo
 * This software is distributed under the terms of the LGPL version 2 and comes WITHOUT ANY WARRANTY.
 * Please read the file COPYING.LIB for further details.
 */

#ifndef __LIBSERIAL_H__
#define __LIBSERIAL_H__

#include <termios.h>
#include <stdint.h>
#include <unistd.h>                                     // ssize_t
#include <sys/ioctl.h>                                  // Constantes indicativas de las líneas serial
#include <fcntl.h>                                      // O_NDELAY

/**-------------------------------------------------------------------------------------------------
 * @brief   Clase Serial: Manejo del puerto serie
 * ------ */
class Serial
{
public:
  /**-------------------------------------------------------------------------------------------------
   * @brief   Tipos enumerados públicos, utilizados por las funciones de la clase
   * ------ */

  /** @brief  Enumeración de la modalidad de lectura (bloqueante o no bloqueante) (Open, SetBlocking) */
  enum EnLockingMode
  {
    Blocking          = 0,                              //!< Las lecturas son bloqueantes (esperan a que haya caracteres que leer)
    NonBlocking       = O_NDELAY                        //!< Las lecturas no son bloqueantes (salen con error si no hay nada que leer)
  };

  /** @brief  Enumeración de los tipos de control de flujo (Open) */
  enum EnFlowControl
  {
    NoFlowCtrl        = 0,
    HardwareFlowCtrl  = CRTSCTS,
    XonXoffInput      = IXOFF,
    XonXoffOutput     = IXON,
    XonXoffBoth       = (IXON | IXOFF)
  };

  /** @brief  Enumeración de los tamaños de carácter (Open) */
  enum EnCharLen
  {
    c5bits            = CS5,
    c6bits            = CS6,
    c7bits            = CS7,
    c8bits            = CS8
  };

  /** @brief  Enumeración de los tipos de paridad (Open) */
  enum EnParity
  {
    NoParity          = 0,
    EvenParity        = PARENB,
    OddParity         = PARODD
  };

  /** @brief  Enumeración de los bits de stop (Open) */
  enum EnStopBits
  {
    stop1bit          = 0,
    stop2bits         = CSTOP
  };

  /** @brief  Estados del buffer (PendingRead, PendingWrite) */
  enum EnPending : int
  {
    PENDING_ERROR = -1,                                 //!< Error: no se ha podido averiguar el estado
    PENDING_EMPTY = 0,                                  //!< No hay datos pendientes de transmitir (buffer vacío)
  };

  /** @brief  Tipos de operación de vaciado de buffer (ClearBuffer) */
  enum EnClearOper
  {
    CLEAR_BUF_IN = 1,                                   //!< Vaciar el buffer de entrada (descartar datos pendientes)
    CLEAR_BUF_OUT,                                      //!< Vaciar el buffer de salida descartando datos pendientes
    FLUSH_BUF_OUT                                       //!< Esperar a que se terminen de enviar los datos pendientes
  };

  /** @brief  Enumeración de las líneas del puerto serie (SetLine, GetLine) */
  enum SerialLine
  {
    LINE_LE   = TIOCM_LE,                               //!<
    LINE_DTR  = TIOCM_DTR,                              //!< DTR (Data Terminal Ready)
    LINE_RTS  = TIOCM_RTS,                              //!< RTS (Ready To Send)
    LINE_ST   = TIOCM_ST,                               //!< ST (Serial Transmit)
    LINE_SR   = TIOCM_SR,                               //!< SR (Serial Receive)
    LINE_CTS  = TIOCM_CTS,                              //!< CTS (Clear To Send)
    LINE_CAR  = TIOCM_CAR,                              //!<
    LINE_RNG  = TIOCM_RNG,                              //!< RNG (Ring)
    LINE_DSR  = TIOCM_DSR                               //!< DSR (Data Send Ready)
  };

  /** @brief  Valores específicos de timeout (Read) */
  enum
  {
    NO_TIMEOUT = 0                                      //!< Lectura sin timeout (@see Read)
  };

  /**-------------------------------------------------------------------------------------------------
   * @brief   Funciones públicas
   * ------ */

  /**
   * @brief   Constructor de la clase - Inicialización de variables
   */
  Serial ();

  /**
   * @brief   Destructor de la clase - Devuelve el puerto serie a su configuración anterior
   */
  ~Serial ();

  /**
   * @brief   Apertura e inicialización del puerto serie
   * @desc    Abre el dispositivo y aplica la configuración solicitada, guardando la anterior.
   * @note    Explicación detallada de los parámetros en la descripción de los tipos enumerados.
   */
  bool                                                  /** @return true: Apertura e inicialización correcta | false: error, o el puerto ya estaba abierto */
  Open (
    const char* devname,                                /** @param  devname     Path al dispositivo serie */
    uint32_t baudrate,                                  /** @param  baudrate    Velocidad (bits por segundo) a establecer */
    EnLockingMode lockmode = NonBlocking,               /** @param  lockmode    Modo de bloqueo en lectura (bloqueante o no bloqueante) */
    EnFlowControl flowcontrol = NoFlowCtrl,             /** @param  flowcontrol Tipo de control de flujo [ = sin control de flujo ] */
    EnCharLen charlen = c8bits,                         /** @param  charlen     Tamaño del carácter [ = 8  bits ] */
    EnParity parity = NoParity,                         /** @param  parity      Control de paridad del carácter [ sin control de paridad ] */
    EnStopBits stopbits = stop1bit                      /** @param  stopbits    Bits de stop (0: 1 stop bit, CSTOP: 2 stop bits) [ = 0 ] */
  );

  /**
   * @brief   Cierre del puerto y restauración de la configuración anterior
   */
  bool                                                  /** @return true: Cierre correcto | false: El puerto ya estaba cerrado */
  Close(
    bool flush = false                                  /** @param flush  true: Esperar al vaciado del buffer de escritura antes de cerrar | false: descartar los buffers */
  );

  /**
   * @brief   Informa de si el puerto está abierto o no
   */
  bool                                                  /** @return true: El puerto está abierto | false: El puerto está cerrado */
  IsOpen()
  { return (handle >= 0); };

  /**
   * @brief   Lectura del puerto serie
   * @desc    En modo Blocking, NO_TIMEOUT espera indefinidamente la llegada de un byte.
   *          En modo NonBlocking, NO_TIMEOUT sale inmediatamente con retorno de 0 bytes si no hay datos que leer.
   */
  ssize_t                                               /** @return -1: error | Bytes leidos */
  Read (
    void* buf,                                          /** @param  buf    Buffer en el que escribir lo leido */
    size_t size,                                        /** @param  size   Bytes a leer */
    uint32_t t_out                                      /** @param  t_out  Timeout en ms, o NO_TIMEOUT. */
  );

  /**
   * @brief   Escritura al puerto serie
   */
  ssize_t                                               /** @return -1: error | Bytes escritos */
  Write (
    const void* buf,                                    /** @param  buf   Buffer con los datos a escribir */
    size_t size                                         /** @param  size  Bytes a escribir */
  );

  /**
   * @brief   Escritura al puerto serie de un sólo byte
   */
  bool                                                  /** @return true: escritura correcta | false: error */
  WriteByte (
    uint8_t byte                                        /** @param  byte  Byte a escribir */
  );

  /**
   * @brief   Comprueba si hay datos pendientes de enviar
   */
  int                                                   /** @return  PENDING_ERROR: Error de lectura | PENDING_EMPTY: Cola de salida vacía | Bytes pendientes de enviar */
  PendingWrite();

  /**
   * @brief   Comprueba si hay datos pendientes de leer
   */
  int                                                   /** @return  PENDING_ERROR: Error de lectura | PENDING_EMPTY: Cola de entrada vacía | Bytes pendientes de recibir */
  PendingRead();

  /**
   * @brief   Espera hasta que se hayan enviado todos los datos
   */
  bool                                                  /** @return true: Operación completada con éxito | false: error */
  WaitSend();

  /**
   * @brief   Vaciar el buffer de entrada, de salida, o ambos
   */
  void                                                  /** @return void */
  ClearBuffer (
    EnClearOper operation                               /** @param  operation   Tipo de operación a realizar: espera o borrado i/o */
  );

  /**
   * @brief   Establecer el estado de las líneas del puerto serie
   */
  bool                                                  /** @return true: Comando completado con éxito | false: error */
  SetLine (
    SerialLine line,                                    /** @param  line    Línea a controlar */
    bool mode                                           /** @param  mode    Valor a establecer (on/off) */
  );

  /**
   * @brief   Recuperar el estado de las líneas del puerto serie
   */
  int                                                   /** @return 1: línea activa | 0: línea inactiva | -1: error */
  GetLine (
    SerialLine line                                     /** @param  line    Línea a controlar */
  );

  /**
   * @brief   Establecer el modo de lectura (bloqueante o no bloqueante)
   */
  bool                                                  /** @return true: Comando completado con éxito | false: error */
  SetBlocking (
    EnLockingMode mode                                  /** @param  mode    Nuevo modo de lectura */
  );

protected:
  /**-------------------------------------------------------------------------------------------------
   * @brief   Estructuras, datos y funciones privados
   * ------ */

  /**
   * @brief   Estructura para asociar valores de baudrate con flags del sistema
   */
  struct Uint2Tcflag
  {
    uint32_t baud;                                      // Baudrate
    tcflag_t flag;                                      // Flag del sistema
  };

  /**
   * @brief   Conversión del parámetro de velocidad del puerto de entero a constante para termios
   */
  tcflag_t                                              /** @return Constante correspondiente o la inmediata inferior */
  GetBaudCode (
    uint32_t baudrate                                   /** @param  baudrate    Valor de velocidad requerido */
  );

  int      handle;                                      //!< File handler
  termios  prev_tio;                                    //!< Configuración anterior del puerto, para restaurarla al cerrar
};

#endif // __LIBSERIAL_H__


src/libserial.cpp
/**
 * libserial
 * Manejo de comunicaciones por puerto serie
 *
 * @file      libserial.cpp
 * @brief     Clase de manejo del puerto serie
 * @author    José Luis Sánchez Arroyo
 * @date      2020.05.22
 * @version   1.7
 *
 * Copyright (c) 2005-2020 José Luis Sánchez Arroyo
 * This software is distributed under the terms of the LGPL version 2 and comes WITHOUT ANY WARRANTY.
 * Please read the file COPYING.LIB for further details.
 */

#include "libserial.h"
#include <libUtility/timer.h>
#include <fcntl.h>
#include <sys/poll.h>
#include <errno.h>

/**-------------------------------------------------------------------------------------------------
 * @brief   Clase Serial: Manejo del puerto serie
 * ------ */

/**
 * @brief     Constructor de la clase - Inicialización de variables
 */
Serial::Serial()
  : handle(-1), prev_tio()
{
}

/**
 * @brief     Destructor de la clase - Devuelve el puerto serie a su configuración anterior
 */
Serial::~Serial()
{
  Close();
}

/**
 * @brief     Apertura e inicialización del puerto serie
 */
bool Serial::Open(const char* devname, uint32_t baudrate, EnLockingMode lockmode, EnFlowControl flowcontrol, EnCharLen charlen, EnParity parity, EnStopBits stopbits)
{
  if (IsOpen() || devname == nullptr)                   // Ya estaba abierto o parámetro erróneo
    return false;

  tcflag_t baud_code = GetBaudCode(baudrate);           // Obtener el código de baudrate correspondiente
  if (baud_code == 0)
    return false;

  handle = open(devname, O_RDWR | O_NOCTTY | lockmode);
  if (handle < 0)                                       // Error de apertura
    return false;

  if (tcgetattr(handle, &prev_tio) < 0)                 // Guardar la anterior configuración del terminal
    return false;

  termios tio = {};                                     // Establecer nuevos parámetros del puerto
  tio.c_iflag = (flowcontrol & (IXON | IXOFF)) | IGNPAR | (parity)? INPCK : 0;
  tio.c_oflag = 0;
  tio.c_cflag = (charlen & CSIZE) | (stopbits & CSTOPB) | CLOCAL | CREAD | (parity? (PARENB | (parity & PARODD)) : 0) | (flowcontrol & CRTSCTS);
  tio.c_lflag = 0;
  tio.c_cc[VTIME] = 0;                                  // Timeout por omisión desactivado
  tio.c_cc[VMIN]  = 1;                                  // Mínimo de caracteres a leer
  cfsetospeed(&tio, baud_code);
  cfsetispeed(&tio, baud_code);
  if (tcflush(handle, TCIFLUSH) < 0 || tcsetattr(handle, TCSANOW, &tio) < 0)
    return false;
  return true;
}

/**
 * @brief     Cierre del puerto y restauración de la configuración anterior
 */
bool Serial::Close(bool flush)
{
  if (!IsOpen())
    return false;
  if (flush)
    tcdrain(handle);
  tcsetattr(handle, TCSANOW, &prev_tio);
  close(handle);
  handle = -1;
  return true;
}

/**
 * @brief     Lectura del puerto serie
 */
ssize_t Serial::Read(void* buf, size_t size, uint32_t t_out)
{
  if (!IsOpen() || buf == nullptr)
    return -1;

  ssize_t rt = 0;
  if (t_out == NO_TIMEOUT)                              // Lectura sin timeout - si Blocking, espera indefinidamente; si NonBlocking, sale al momento.
  {
    rt = read(handle, buf, size);
    if (rt < 0 && errno == EAGAIN)                      // Lectura no bloqueante: esto no es un error, es que no hay nada que leer
      return 0;
    return rt;
  }

  pollfd p_list;
  p_list.fd = handle;
  p_list.events = POLLIN;
  int err;
  Timer timer;
  timer.SetAlarm(t_out);
  uint8_t* ptr = reinterpret_cast<uint8_t*>(buf);       // Necesario para poder hacer aritmética de punteros
  for (rt = 0; rt < static_cast<ssize_t>(size); )
  {
    do
      err = poll(&p_list, 1, timer.GetRemain());
    while (err < 0 && errno == EINTR);                  // Continuar a la espera si se recibe EINTR
    if (err <= 0 || timer.IsExpired())                  // Salida con error o timeout
      break;
    err = read(handle, &ptr[rt], size - rt);            // Se supone que esto leerá algo...
    if (err < 0)                                        // Error de lectura
      break;
    rt += err;
  }
  return rt;
}

/**
 * @brief     Escritura al puerto serie
 */
ssize_t Serial::Write(const void* buf, size_t size)
{
  if (!IsOpen() || buf == nullptr)
    return -1;
  ssize_t bytes = write(handle, buf, size);
  return (bytes >= 0)? bytes : -1;
}

/**
 * @brief     Escritura al puerto serie de un sólo byte
 */
bool Serial::WriteByte(uint8_t byte)
{
  if (!IsOpen())
    return false;
  return (write(handle, &byte, 1) == 1);
}

/**
 * @brief     Comprueba si hay datos pendientes de enviar
 */
int Serial::PendingWrite()
{
  int pending, lsr;
  if (ioctl(handle, TIOCSERGETLSR, &lsr) < 0)           // lectura de line status register
    return PENDING_ERROR;
  if (lsr & TIOCSER_TEMT)                               // FIFO y shift register vacíos
    return PENDING_EMPTY;
  if (ioctl(handle, TIOCOUTQ, &pending) < 0)
    return PENDING_ERROR;
  return pending? pending : 1;
}

/**
 * @brief     Comprueba si hay datos pendientes de leer
 */
int Serial::PendingRead()
{
  int pending;
  if (ioctl(handle, TIOCINQ, &pending) < 0)
    return PENDING_ERROR;
  return pending;
}

/**
 * @brief     Espera hasta que se hayan enviado todos los datos
 */
bool Serial::WaitSend()
{
  while (true)
    switch (PendingWrite())
    {
      case PENDING_ERROR:
        return false;
      case PENDING_EMPTY:
        return true;
      default:
        Msleep(1);
    }
}

/**
 * @brief     Vaciar el buffer de entrada, de salida, o ambos
 */
void Serial::ClearBuffer(EnClearOper operation)
{
  if (IsOpen())
  {
    switch (operation)
    {
      case CLEAR_BUF_IN:                                // Borrar buffer de entrada
        tcflush(handle, TCIFLUSH);
        break;
      case CLEAR_BUF_OUT:                               // Borrar buffer de salida
        tcflush(handle, TCOFLUSH);
        break;
      case FLUSH_BUF_OUT:                               // Esperar vaciado del buffer de salida
        tcdrain(handle);                                // El kernel ha volcado los datos a la UART
        break;
    }
  }
}

/**
 * @brief     Establecer el estado de las líneas del puerto serie
 */
bool Serial::SetLine(SerialLine line, bool mode)
{
  if (!IsOpen())
    return false;
  unsigned flags;
  if (ioctl(handle, TIOCMGET, &flags) < 0)
    return false;

  if (mode)
    flags |= line;
  else
    flags &= ~line;
  return (ioctl(handle, TIOCMSET, &flags) >= 0);
}

/**
 * @brief   Recuperar el estado de las líneas del puerto serie
 */
int Serial::GetLine(SerialLine line)
{
  if (!IsOpen())
    return -1;
  unsigned flags;
  if (ioctl(handle, TIOCMGET, &flags) < 0)
    return -1;
  if (flags & line)
    return 1;
  return 0;
}
/**
 * @brief   Establecer el modo de lectura (bloqueante o no bloqueante)
 */
bool Serial::SetBlocking (EnLockingMode mode)
{
  if (!IsOpen())
    return false;
  int flags = fcntl(handle, F_GETFL, 0);
  if (flags == -1)
    return false;
  if (mode)
    flags |= O_NDELAY;
  else
    flags &= ~O_NDELAY;
  return (fcntl(handle, F_SETFL, flags) >= 0);
}

/**------------------------------------------
 * @brief     Funciones privadas
 * ------ */

/**
 * @brief     Conversión del parámetro de velocidad del puerto de entero a constante válida para Init.
 */
tcflag_t Serial::GetBaudCode(uint32_t baudrate)
{
  static Uint2Tcflag uint2tcflag[] =                    //!< Tabla de equivalencias de flags y valores de bps
  {
    { 4000000, B4000000 },
    { 3500000, B3500000 },
    { 3000000, B3000000 },
    { 2500000, B2500000 },
    { 2000000, B2000000 },
    { 1500000, B1500000 },
    { 1152000, B1152000 },
    { 1000000, B1000000 },
    {  921600, B921600  },
    {  576000, B576000  },
    {  500000, B500000  },
    {  460800, B460800  },
    {  230400, B230400  },
    {  115200, B115200  },
    {   57600, B57600   },
    {   38400, B38400   },
    {   19200, B19200   },
    {    9600, B9600    },
    {    4800, B4800    },
    {    2400, B2400    },
    {    1800, B1800    },
    {    1200, B1200    },
    {     600, B600     },
    {     300, B300     },
    {     200, B200     },
    {     150, B150     },
    {     134, B134     },
    {     110, B110     },
    {      75, B75      },
    {      50, B50      },
  };

  for (auto iter : uint2tcflag)
    if (iter.baud <= baudrate)
      return iter.flag;
  return 0;                                           // fallback
}
